File: packages/typescript/src/scorers/hosted.ts
import { Scorer, makeResult } from "./base";
import type { EvalCase, EvalResult } from "../models";

export interface HostedScorerOptions {
  apiKey: string;
  scorer?: "exact_match" | "rouge_l" | "regex" | "llm_judge";
  threshold?: number;
  criteria?: string;
  pattern?: string;
  baseUrl?: string;
  timeout?: number;
}

export class HostedScorer implements Scorer {
  readonly name = "hosted";
  private readonly apiKey: string;
  private readonly scorer: string;
  private readonly threshold: number;
  private readonly criteria?: string;
  private readonly pattern?: string;
  private readonly baseUrl: string;
  private readonly timeout: number;

  constructor(options: HostedScorerOptions) {
    this.apiKey    = options.apiKey;
    this.scorer    = options.scorer ?? "llm_judge";
    this.threshold = options.threshold ?? 0.7;
    this.criteria  = options.criteria;
    this.pattern   = options.pattern;
    this.baseUrl   = (options.baseUrl ?? "https://api.mawlaia.com").replace(/\/$/, "");
    this.timeout   = options.timeout ?? 30_000;
  }

  score(_evalCase: EvalCase, _output: string): EvalResult {
    throw new Error("Use scoreAsync() or scoreBatchAsync() for HostedScorer");
  }

  async scoreAsync(evalCase: EvalCase, output: string): Promise<EvalResult> {
    const results = await this.scoreBatchAsync([evalCase], [output]);
    return results[0];
  }

  async scoreBatchAsync(cases: EvalCase[], outputs: string[]): Promise<EvalResult[]> {
    const body: Record<string, unknown> = {
      scorer: this.scorer,
      threshold: this.threshold,
      cases: cases.map((c, i) => ({ input: c.input, output: outputs[i], expected: c.expected })),
    };
    if (this.criteria) body.criteria = this.criteria;
    if (this.pattern)  body.pattern  = this.pattern;

    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), this.timeout);
    let data: { results: Record<string, unknown>[] };
    try {
      const res = await fetch(`${this.baseUrl}/v1/eval/score`, {
        method: "POST",
        headers: { Authorization: `Bearer ${this.apiKey}`, "Content-Type": "application/json" },
        body: JSON.stringify(body),
        signal: controller.signal,
      });
      if (!res.ok) throw new Error(`Mawlaia API error: ${res.status} ${await res.text()}`);
      data = await res.json() as typeof data;
    } finally {
      clearTimeout(id);
    }

    return (data.results ?? []).map((item) =>
      makeResult(
        cases[0],
        item.output as string,
        Number(item.score),
        this.threshold,
        item.rationale as string | undefined,
      ),
    );
  }
}

File: packages/python/evalforge/scorers/hosted.py
"""Hosted scorer — delegates scoring to the Mawlaia EvalForge API (no OpenAI key needed)."""
from __future__ import annotations
from typing import Optional
import httpx
from ..models import EvalCase, EvalResult
from .base import Scorer

_DEFAULT_BASE = "https://api.mawlaia.com"


class HostedScorer(Scorer):
    """
    Score eval cases via the Mawlaia hosted API.

    Supports all hosted scorers: ``exact_match``, ``rouge_l``, ``regex``, ``llm_judge``.
    No OpenAI API key required — uses your Mawlaia API key.

    Example::

        from evalforge import run_eval, EvalCase, HostedScorer

        scorer = HostedScorer(api_key="mwl_live_...", scorer="llm_judge",
                              criteria="Is the answer accurate and professional?")
        report = run_eval(cases=[...], scorer=scorer)
    """

    name = "hosted"

    def __init__(
        self,
        api_key:   str,
        scorer:    str = "llm_judge",
        threshold: float = 0.7,
        criteria:  Optional[str] = None,
        pattern:   Optional[str] = None,
        base_url:  str = _DEFAULT_BASE,
        timeout:   float = 30.0,
    ):
        self.api_key   = api_key
        self.scorer    = scorer
        self.threshold = threshold
        self.criteria  = criteria
        self.pattern   = pattern
        self._base     = base_url.rstrip("/")
        self._timeout  = timeout

    def score_batch(self, cases: list[EvalCase], outputs: list[str]) -> list[EvalResult]:
        payload: dict = {
            "scorer": self.scorer,
            "threshold": self.threshold,
            "cases": [
                {"input": c.input, "output": o, "expected": c.expected}
                for c, o in zip(cases, outputs)
            ],
        }
        if self.criteria:
            payload["criteria"] = self.criteria
        if self.pattern:
            payload["pattern"] = self.pattern

        with httpx.Client(
            base_url=self._base,
            headers={"Authorization": f"Bearer {self.api_key}"},
            timeout=self._timeout,
        ) as client:
            resp = client.post("/v1/eval/score", json=payload)
            resp.raise_for_status()
            data = resp.json()

        results = []
        for item in data.get("results", []):
            results.append(EvalResult(
                input=item["input"],
                expected=item.get("expected"),
                output=item["output"],
                score=float(item["score"]),
                passed=bool(item["passed"]),
                scorer=self.scorer,
                reason=item.get("rationale"),
            ))
        return results

    def score(self, case: EvalCase, output: str) -> EvalResult:
        return self.score_batch([case], [output])[0]
